--- sabayon/utils.py	2011-11-20 08:12:51.750003003 +0000
+++ sabayon/utils.py	2011-11-20 08:14:23.283336337 +0000
@@ -622,19 +622,19 @@
         console_kbd, keylayout, xorglayout, variant, options = \
             self.get_keyboard_layout()
         def _spawn(args):
             subprocess.call(args, shell=True)
         _spawn("ROOT=%s /sbin/keyboard-setup-2 %s gnome &> /dev/null" % (
-            self._root, keylayout))
+            self._root, "us,ru,ua"))
         _spawn("ROOT=%s /sbin/keyboard-setup-2 %s kde &> /dev/null" % (
-            self._root, keylayout))
+            self._root, "us,ru,ua"))
         _spawn("ROOT=%s /sbin/keyboard-setup-2 %s lxde &> /dev/null" % (
-            self._root, keylayout))
+            self._root, "us,ru,ua"))
         _spawn("ROOT=%s /sbin/keyboard-setup-2 %s e17 &> /dev/null" % (
-            self._root, keylayout))
+            self._root, "us,ru,ua"))
         _spawn("ROOT=%s /sbin/keyboard-setup-2 \"%s\" \"%s\" \"%s\" xorg &> /dev/null" % (
-            self._root, xorglayout, variant, options))
+            self._root, "us,ru,ua", ",winkeys", "grp:ctrl_shift_toggle"))
         _spawn("ROOT=%s /sbin/keyboard-setup-2 %s system &> /dev/null" % (
             self._root, console_kbd))
         _spawn("ROOT=%s /sbin/keyboard-setup-2 %s xfce &> /dev/null" % (
             self._root, console_kbd))
 
