Only in patched2/: 04_deprecated_GTK.patch
diff -U3 -r orig//auth-dialog/gnome-two-password-dialog.c patched2//auth-dialog/gnome-two-password-dialog.c
--- orig//auth-dialog/gnome-two-password-dialog.c	2011-01-31 18:13:05.000000000 +0000
+++ patched2//auth-dialog/gnome-two-password-dialog.c	2012-01-16 21:54:41.041693001 +0000
@@ -158,15 +158,15 @@
 
 	password_dialog = GNOME_TWO_PASSWORD_DIALOG (callback_data);
 
-	if (GTK_WIDGET_VISIBLE (password_dialog->details->username_entry) &&
+	if (gtk_widget_get_visible (password_dialog->details->username_entry) &&
 	    !password_dialog->details->readonly_username) {
 		gtk_widget_grab_focus (password_dialog->details->username_entry);
-	} else if (GTK_WIDGET_VISIBLE (password_dialog->details->domain_entry) &&
+	} else if (gtk_widget_get_visible (password_dialog->details->domain_entry) &&
 		   !password_dialog->details->readonly_domain) {
 		gtk_widget_grab_focus (password_dialog->details->domain_entry);
-	} else if (GTK_WIDGET_VISIBLE (password_dialog->details->password_entry)) {
+	} else if (gtk_widget_get_visible (password_dialog->details->password_entry)) {
 		gtk_widget_grab_focus (password_dialog->details->password_entry);
-	} else if (GTK_WIDGET_VISIBLE (password_dialog->details->password_entry_secondary)) {
+	} else if (gtk_widget_get_visible (password_dialog->details->password_entry_secondary)) {
 		gtk_widget_grab_focus (password_dialog->details->password_entry_secondary);
 	}
 }
@@ -259,14 +259,14 @@
 
 	password_dialog = GNOME_TWO_PASSWORD_DIALOG (dialog);
 	
-	if (GTK_WIDGET_VISIBLE (password_dialog->details->domain_entry) &&
-	    GTK_WIDGET_SENSITIVE (password_dialog->details->domain_entry))
+	if (gtk_widget_get_visible (password_dialog->details->domain_entry) &&
+	    gtk_widget_get_sensitive (password_dialog->details->domain_entry))
 		gtk_widget_grab_focus (password_dialog->details->domain_entry);
-	else if (GTK_WIDGET_VISIBLE (password_dialog->details->password_entry) &&
-		 GTK_WIDGET_SENSITIVE (password_dialog->details->password_entry))
+	else if (gtk_widget_get_visible (password_dialog->details->password_entry) &&
+		 gtk_widget_get_sensitive (password_dialog->details->password_entry))
 		gtk_widget_grab_focus (password_dialog->details->password_entry);
-	else if (GTK_WIDGET_VISIBLE (password_dialog->details->password_entry_secondary) &&
-		 GTK_WIDGET_SENSITIVE (password_dialog->details->password_entry_secondary))
+	else if (gtk_widget_get_visible (password_dialog->details->password_entry_secondary) &&
+		 gtk_widget_get_sensitive (password_dialog->details->password_entry_secondary))
 		gtk_widget_grab_focus (password_dialog->details->password_entry_secondary);
 }
 
@@ -277,11 +277,11 @@
 
 	password_dialog = GNOME_TWO_PASSWORD_DIALOG (dialog);
 	
-	if (GTK_WIDGET_VISIBLE (password_dialog->details->password_entry) &&
-	    GTK_WIDGET_SENSITIVE (password_dialog->details->password_entry))
+	if (gtk_widget_get_visible (password_dialog->details->password_entry) &&
+	    gtk_widget_get_sensitive (password_dialog->details->password_entry))
 		gtk_widget_grab_focus (password_dialog->details->password_entry);
-	else if (GTK_WIDGET_VISIBLE (password_dialog->details->password_entry_secondary) &&
-		 GTK_WIDGET_SENSITIVE (password_dialog->details->password_entry_secondary))
+	else if (gtk_widget_get_visible (password_dialog->details->password_entry_secondary) &&
+		 gtk_widget_get_sensitive (password_dialog->details->password_entry_secondary))
 		gtk_widget_grab_focus (password_dialog->details->password_entry_secondary);
 }
 
Only in patched2/: gtk_widget_get_visible.patch
