diff -dPNur p7zip_4.42/7zip/Bundles/Un7Zip/Un7Zip.cpp p7zip_4.42-new/7zip/Bundles/Un7Zip/Un7Zip.cpp
--- p7zip_4.42/7zip/Bundles/Un7Zip/Un7Zip.cpp	2006-06-28 04:12:35.000000000 +0500
+++ p7zip_4.42-new/7zip/Bundles/Un7Zip/Un7Zip.cpp	2006-06-28 04:25:56.000000000 +0500
@@ -117,6 +117,7 @@
           v1, v2,
           wildcardCensorHead, 
           true, 
+	  false,
           passwordEnabled, 
           password);
       if (result != S_OK)
