--- trunk/KDE/kdebase/workspace/kwin/tabbox/tabboxconfig.cpp	2010/03/19 10:09:52	1105143
+++ trunk/KDE/kdebase/workspace/kwin/tabbox/tabboxconfig.cpp	2010/03/19 10:15:56	1105144
@@ -41,6 +41,7 @@
             , minHeight( TabBoxConfig::defaultMinHeight() )
             , layoutName( TabBoxConfig::defaultLayoutName() )
             , selectedItemLayoutName( TabBoxConfig::defaultSelectedItemLayoutName() )
+            , showDesktop( TabBoxConfig::defaultShowDesktop() )
             {
             }
         ~TabBoxConfigPrivate()
@@ -60,6 +61,7 @@
         int minHeight;
         QString layoutName;
         QString selectedItemLayoutName;
+        bool showDesktop;
     };
 
 TabBoxConfig::TabBoxConfig()
@@ -78,6 +80,7 @@
     d->highlightWindows = object.isHighlightWindows();
     d->showOutline = object.isShowOutline();
     d->tabBoxMode = object.tabBoxMode();
+    d->showDesktop = object.isShowDesktop();
     d->layout = object.layout();
     d->clientListMode = object.clientListMode();
     d->clientSwitchingMode = object.clientSwitchingMode();
@@ -220,5 +223,15 @@
     d->selectedItemLayoutName = name;
     }
 
+bool TabBoxConfig::isShowDesktop() const
+    {
+    return d->showDesktop;
+    }
+
+void TabBoxConfig::setShowDesktop( bool show )
+    {
+    d->showDesktop = show;
+    }
+
 } // namespace TabBox
 } // namespace KWin
