--- trunk/KDE/kdebase/workspace/kwin/tabbox.h	2010/03/19 10:09:52	1105143
+++ trunk/KDE/kdebase/workspace/kwin/tabbox.h	2010/03/19 10:15:56	1105144
@@ -55,6 +55,7 @@
         virtual int numberOfDesktops() const;
         virtual TabBoxClientList stackingOrder() const;
         virtual TabBoxClient* clientToAddToList( TabBoxClient* client, int desktop, bool allDesktops ) const;
+        virtual TabBoxClient* desktopClient() const;
     };
 
 class TabBoxClientImpl : public TabBoxClient
