--- trunk/KDE/kdebase/workspace/kwin/tabbox/clientmodel.cpp	2010/03/19 10:09:52	1105143
+++ trunk/KDE/kdebase/workspace/kwin/tabbox/clientmodel.cpp	2010/03/19 10:15:56	1105144
@@ -226,6 +226,12 @@
             break;
             }
         }
+    if( tabBox->config().isShowDesktop() )
+        {
+        TabBoxClient* desktopClient = tabBox->desktopClient();
+        if( desktopClient )
+            m_clientList.append( desktopClient );
+        }
     reset();
     }
 
