--- trunk/KDE/kdebase/workspace/kwin/tabbox/tabboxhandler.h	2010/03/19 10:09:52	1105143
+++ trunk/KDE/kdebase/workspace/kwin/tabbox/tabboxhandler.h	2010/03/19 10:15:56	1105144
@@ -157,6 +157,10 @@
         * @return The client to be included in the list or NULL if it isn't to be included
         */
         virtual TabBoxClient* clientToAddToList( TabBoxClient* client, int desktop, bool allDesktops ) const = 0;
+        /**
+        * @return The first desktop window in the stacking order.
+        */
+        virtual TabBoxClient* desktopClient() const = 0;
 
         /**
         * @return The currently used TabBoxConfig
