diff -dPNur xmms-1.2.10/Input/mpg123/id3_frame_content.c xmms-1.2.10-new/Input/mpg123/id3_frame_content.c
--- xmms-1.2.10/Input/mpg123/id3_frame_content.c	2006-11-23 10:09:31.000000000 +0400
+++ xmms-1.2.10-new/Input/mpg123/id3_frame_content.c	2006-11-23 10:10:28.000000000 +0400
@@ -48,6 +48,8 @@
 	text_beg = text = id3_string_decode(ID3_TEXT_FRAME_ENCODING(frame),
 					    ID3_TEXT_FRAME_PTR(frame));
 
+	if (!text_beg) return NULL;
+
 	/*
 	 * If content is just plain text, return it.
 	 */
